### DISCUSSION TITLE
How subscribe mqtt with scripting

### DISCUSSION THREAD

**Discussion by rice103** [arendst/Tasmota#10272] [Q&A] [answered]
For rules there is different example about to subscribe mqtt.
But on help you can know only if is possibile also with script but not how:
```
Subscribe, Unsubscribe
#define SUPPORT_MQTT_EVENT
subscribe and unsubscribe commands are supported. In contrast to rules, no event is generated but the event name specifies a variable defined in D section and this variable is automatically set on transmission of the subscribed item
within a script the subscribe cmd must be send with +> instead of =>
the MQTT decoder may be configured for more space in user config overwrite by
#define MQTT_EVENT_MSIZE xxx (default is 256)
#define MQTT_EVENT_JSIZE xxx (default is 400)
```
Someone could help me?

RULE:
Rule1 ON mqtt#connected DO Subscribe DnTemp, tele/watertemperature/SENSOR, DS18B20-2.Temperature ENDON

**Comment by gemu2015** [comment]
1. math operators in JSON Messages are not allowed. (DS18B20-2) in Tasmota you may configure this by SetOption64 1  this replaces the - char with underscore _
2. #define MQTT_EVENT_JSIZE xxx (default is 400) is no longer needed since the JSON library has changed.
3. in contrast to rules all script JSON strings and variables are case sensitive
4. in stead of an event you specify a script variable. this is automatically set on arrival of the subscription and an >E event is generated.

**Reply by rice103** [reply]
But please leave me an example

**Reply by rice103** [reply]
DS18B20-2 is second DS18B20 on tasmota publisher

**Comment by gemu2015** [comment, marked as answer]
Yes i know, you have to do setoption64 1 on the publisher.
its a bit tricky to publish inside a script because MQTT must be up and running
upon restart of a script with running MQTT section in >B is executed because MQTT already running
upon restart of the device the >B section is omitted because MQTT is not yet running
so the publish is executed in >S when MQTT starts

\>D
DnTemp=0

\>B
if mqtts>0
then
=#publish
endif

#publish
+>Subscribe DnTemp, tele/watertemperature/SENSOR, DS18B20_2.Temperature 

\>S
if mqttc>0
then
=#publish
endif

\>E
print %DnTemp%

what do you want to accomplish ?
i assume you want to get a sensor from another Tasmota ?
this imho can be better solved by global variables.
this even works without any MQTT server.
just to give an example
#define USE_SCRIPT_GLOBVARS

in the "sender"
\>D
g:DnTemp=0

\>T
DnTemp=DS18B20_2#Temperature 

that's all

in the "receiver" or as many as you want receivers
\>D
g:DnTemp=0

\>G
if upd[DnTemp]>0
then
print %DnTemp%
endif

**Reply by rice103** [reply]
WOW wonderful,
i don't know globalvars but you're right, i'm using temperature from device to control thermostat on another.
I like more mqtt than udp-multicast but it's another good solution to interchange value.
Thanks a lot!
Francesco

**Reply by rice103** [reply]
Why??
somebody knows?
![image](https://user-images.githubusercontent.com/35748699/103179262-5f826580-488a-11eb-9477-3312c1402092.png)

**Reply by barbudor** [reply]
@rice103 The problem on the script editor should have been solved in the latest development version
Please upgrade to 9.2.0.1+

**Reply by rice103** [reply]
Ok thanks

**Comment by letshin** [comment]
Hi @gemu2015 

I tried this example and also the one in Issue #16943.

```
>D
DnTemp=0

>B
if mqtts>0
then
=#publish
endif

#publish
+>Subscribe DnTemp, tele/watertemperature/SENSOR, msg

>S
if mqttc>0
then
=#publish
endif

>E
print %DnTemp%
```
When I send the MQTT topic I don't seem to get a response:
![image](https://user-images.githubusercontent.com/40326519/233743782-db60400c-b497-443f-844a-c95f41ee8e84.png)

This is what comes out of the console:
```
00:00:16.314 MQT: stat/tasmota_AAC3FA/RESULT = {"Subscribe":"DnTemp,tele/watertemperature/SENSOR/#,msg"}
00:00:16.317 0.00
00:00:17.454 MQT: stat/tasmota_AAC3FA/RESULT = {"Subscribe":"DnTemp,tele/watertemperature/SENSOR/#,msg"}
00:00:17.457 0.00
00:01:40.598 MQT: stat/tasmota_AAC3FA/RESULT = {"POWER":"OFF"}
00:01:40.602 MQT: stat/tasmota_AAC3FA/POWER = OFF
00:01:45.892 MQT: stat/tasmota_AAC3FA/RESULT = {"POWER":"OFF"}
00:01:45.899 MQT: stat/tasmota_AAC3FA/POWER = OFF
00:05:13.450 MQT: tele/tasmota_AAC3FA/STATE = {"Time":"1970-01-01T00:05:13","Uptime":"0T00:05:13","UptimeSec":313,"Heap":22,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"OFF","Wifi":{"AP":1,"SSId":"A4AB WiFi","BSSId":"EA:9F:80:DA:50:8A","Channel":1,"Mode":"11n","RSSI":50,"Signal":-75,"LinkCount":1,"Downtime":"0T00:00:05"}}
```

In this case I am trying to get the WiFi Kit 8 to display text received from a MQTT topic. Am I misunderstanding how to do this please?

Edit: When I add
```
>S 
dt [s1l1c1] %DnTemp%
```
The text shows up on the OLED, but it is not printed out to console. Is this a bug or a feature?

**Reply by gemu2015** [reply]
at the time of writing the above code worked.
you may compile with
#define DEBUG_MQTT_EVENT
and see what is going on.

the Variable is set properly since you can print it out.
obviously if not yet set when section >E is called.

as a workaround try this

```
>S
if upd[DnTemp]>0 {
print %DnTemp%
}
```
